src/components/structures/MatrixChat.tsx

```diff
@@ -1868,42 +1868,31 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
             this.accountPasswordTimer = null;
         }, 60 * 5 * 1000);
 
-        // Wait for the client to be logged in (but not started)
-        // which is enough to ask the server about account data.
-        const loggedIn = new Promise(resolve => {
-            const actionHandlerRef = dis.register(payload => {
-                if (payload.action !== "on_logged_in") {
-                    return;
-                }
-                dis.unregister(actionHandlerRef);
-                resolve();
-            });
-        });
-
-        // Create and start the client in the background
-        const setLoggedInPromise = Lifecycle.setLoggedIn(credentials);
-        await loggedIn;
+        // Create and start the client
+        await Lifecycle.setLoggedIn(credentials);
 
         const cli = MatrixClientPeg.get();
-        // We're checking `isCryptoAvailable` here instead of `isCryptoEnabled`
-        // because the client hasn't been started yet.
-        const cryptoAvailable = isCryptoAvailable();
-        if (!cryptoAvailable) {
+        const cryptoEnabled = cli.isCryptoEnabled();
+        if (!cryptoEnabled) {
             this.onLoggedIn();
         }
 
+        const promisesList = [this.firstSyncPromise.promise];
+        if (cryptoEnabled) {
+            // wait for the client to finish downloading cross-signing keys for us so we
+            // know whether or not we have keys set up on this account
+            promisesList.push(cli.downloadKeys([cli.getUserId()]));
+        }
+
         this.setState({ pendingInitialSync: true });
```

Comment:
this probably needs moving up now that we block on things sooner

Reply:
Hmm, we are waiting for more things now but `pendingInitialSync` is still for the first sync rather than the rest of the login stuff, so I think this is still probably in the right place, although in reality the difference is pretty minimal. I've added a comment.